### DISCUSSION TITLE
x-bind to $root

### DISCUSSION THREAD

**Discussion by manuelmeister** [alpinejs/alpine#3784] [1. Help] [answered]
It would be really nice to have a way to x-bind to the reusable component without declaring x-bind besides x-data:

```js
Alpine.data("link", () => ({
    clickable: false,
    click(event) {
        if (!this.clickable) {
            event.preventDefault();
        }
    },
    root: {
        ["x-on:click.capture"]: function (event) {
            this.click(event);
        },
        ["x-on:keydown.enter"]: function (event) {
            this.click(event);
        },
        ["x-on:keydown.shift"]: function () {
            this.clickable = true;
        },
        ["x-on:keyup"]: function () {
            this.clickable = false;
        },
    }
}));
```

How we have to do it now:
```html
 <a x-data="link" x-bind="root">Read more</a>
```

How I would love to do it:
```html
<a x-data="link">Read more</a>
```

**Comment by ekwoka** [comment]
There already is:

https://alpinejs.dev/globals/alpine-bind

**Comment by ekwoka** [comment, marked as answer]
oh, you mean, have like a property on the x-data that auto binds stuff.

So you can add to your data

```js
init() {
  Alpine.bind(this.$root, this.root)
}
```

And it would do that.

Also, as a style thing, the normal way to define a method on an object is

```js
["x-on:click.capture"](event) {
    this.click(event);
}
```

not with a function expression

**Reply by manuelmeister** [reply]
Cool. Do you think such a field would make sense?
For instance `$root`?

**Reply by ekwoka** [reply]
$root already exists

So reusing it wouldn't make any sense.

I think just calling alpine.bind on the element is simple enough already

**Reply by manuelmeister** [reply]
So this has no negative side effects? This way seems "hackish".

**Reply by ekwoka** [reply]
Not at all, it's basically what `x-bind` does under the hood.

It just evaluates the expression, if it finds an object it calls bind on that element with the object.

It does a few other things to handle certain circumstances, but since you're not calling it in an effect, you don't need to worry about those, you can just call it directly with no issues.

It would actually be more performant by quite a bit for you to call `Alpine.bind` like this than to have the `x-bind` on the element.

**Reply by y1n0** [reply]
I second this. I mean it feels so natural to expect the properties to auto-bind themselves without adding a "root" key.
I wristled with this code for a long time trying to figure out why i did not work.
```js
   # does not work
    Alpine.data('comp', () => ({
        init() { console.log('init') },
        ['@click']() { console.log('click') },
        ['@keyup'](e) { console.log('keyup') }
    }));
```

**Reply by y1n0** [reply]
I still somehow can not seem to make it work.. in the example from the [documentation](https://alpinejs.dev/globals/alpine-bind), is that button supposed to be a descendant of an element on which `x-data` or x-init were already called? 

```html
<div x-data>
  <button x-bind="SomeButton"></button>
</div>
```
just adding x-bind to it and calling `Alpine.bind('SomeButton', () => {})` does not work.

**Reply by ekwoka** [reply]
It does though

https://awesomealpine.com/play/?coreplugins=0&html=PGRpdgogIHgtZGF0YQogIHgtYmluZD0iY29vbHN0dWZmIgo-CiAgaGVsbG8KPC9kaXY-Cg&script=QWxwaW5lLmJpbmQoJ2Nvb2xzdHVmZicsIHsKICBzdHlsZTogJ2NvbG9yOndoaXRlO2JhY2tncm91bmQ6YmxhY2s7JywKfSk7Cg

**Reply by SimoTod** [reply]
> I second this. I mean it feels so natural to expect the properties to auto-bind themselves without adding a "root" key.
> I wristled with this code for a long time trying to figure out why i did not work.
> ```js
>    # does not work
>     Alpine.data('comp', () => ({
>         init() { console.log('init') },
>         ['@click']() { console.log('click') },
>         ['@keyup'](e) { console.log('keyup') }
>     }));
> ```

How do you differentiate between a data property in your component vs a property that you want to bind to your component? Let's not say that by comparing the name to existing directives because you can have custom prefixes, custom directives, arbitrary modifiers, you may want to bind a standard attribute, etc.

**Reply by ekwoka** [reply]
Yeah, I think registering bindings with `Alpine.bind` and using them or applying them in the `init` with `Alpine.bind` is simple enough in user land without needing to introduce magic that could make things less clear and reduce the posibilities of the future..

**Reply by y1n0** [reply]
> It does though
> 
> https://awesomealpine.com/play/?coreplugins=0&html=PGRpdgogIHgtZGF0YQogIHgtYmluZD0iY29vbHN0dWZmIgo-CiAgaGVsbG8KPC9kaXY-Cg&script=QWxwaW5lLmJpbmQoJ2Nvb2xzdHVmZicsIHsKICBzdHlsZTogJ2NvbG9yOndoaXRlO2JhY2tncm91bmQ6YmxhY2s7JywKfSk7Cg

Thanks for providing the sandbox. So `x-data` is needed for `x-bind` to work. Maybe that should be mentioned in x-bind documentation because it reads like x-bind is enough by itself, thus my initial confusion. 
Edit: I get it `x-data` or `x-init` are what actually initializes the component so it is implied they must have been called somewhere up in the dom tree, but still I think I would not hurt to mention it again for clarity

**Reply by ekwoka** [reply]
This comes up a bit, but idk if adding that note to every single page on the docs is a solution...

maybe though
